Add contrast to text drawn in player colors

obrakmann opened the pull request:
- End game screen
- Observer stats
- Observer shroud drop down
- Ingame and lobby chat
- Actor tooltips
- Replay browser player list

Fixes #7592.

GraionDilach commented:
:+1:

Mailaender commented:
@RobotCaleb Does this help colorblind gamers?

GraionDilach commented:
Probably only as much as ordinary people - color blindness affects only hues from what I know.

Mailaender commented:
Is it on purpose that the "shadow" is actually drawn above the text?

obrakmann commented:
It's not drawn above the text, but the shadow is indeed on the top-left. I found it made things more readable compared to the shadow being on the bottom-right. You can compare yourself by replacing the `-1` on [this line](https://github.com/OpenRA/OpenRA/pull/11693/files#diff-af8247f6d0d31e2b5135557ba1aac4a8R99) with `1`. I'm fine with changing that if that's just due to my quirky eyes.

pchote commented:
Could someone post a screenshot of these new shadows?

RobotCaleb commented:
I agree with pchote. Please provide me with some media to review. I'm not in a state to build and test myself. (This sort of PR should probably always include reference.)

Before and after would be great.

obrakmann commented:
Screenshots:

Bleed:
![screenshot from 2016-07-25 19 21 36](https://cloud.githubusercontent.com/assets/4331210/17110485/7d87de9e-529d-11e6-86db-ac60dfa1416f.png)
![screenshot from 2016-07-25 19 21 43](https://cloud.githubusercontent.com/assets/4331210/17110484/7d845706-529d-11e6-8ae3-3f9300cb6464.png)

With shadows on the top-left of the letters (current state of this branch):

![screenshot from 2016-07-25 19 14 42](https://cloud.githubusercontent.com/assets/4331210/17110500/9a3c7b62-529d-11e6-8db5-e6ec81d60365.png)
![screenshot from 2016-07-25 19 14 55](https://cloud.githubusercontent.com/assets/4331210/17110501/9a3ccb9e-529d-11e6-9bbf-47e3aaf8598d.png)

With shadows on the bottom-right of the letters:
![screenshot from 2016-07-25 19 17 56](https://cloud.githubusercontent.com/assets/4331210/17110521/b1cabf28-529d-11e6-9a51-3b6803b3440c.png)
![screenshot from 2016-07-25 19 17 34](https://cloud.githubusercontent.com/assets/4331210/17110522/b1cbad0c-529d-11e6-8c7c-add4ccddd665.png)

The three smaller ones side-by-side:
![screenshot from 2016-07-25 19 21 43](https://cloud.githubusercontent.com/assets/4331210/17110484/7d845706-529d-11e6-8ae3-3f9300cb6464.png) ![screenshot from 2016-07-25 19 14 55](https://cloud.githubusercontent.com/assets/4331210/17110501/9a3ccb9e-529d-11e6-9bbf-47e3aaf8598d.png) ![screenshot from 2016-07-25 19 17 34](https://cloud.githubusercontent.com/assets/4331210/17110522/b1cbad0c-529d-11e6-8c7c-add4ccddd665.png)

pchote commented:
Bottom right shadow looks much better, IMO.

pchote commented:
For consistency you might want to add shadow to the other labels too (country name, score, etc)

obrakmann commented:
Done.

RobotCaleb commented:
Thanks for providing the screenshots.

In the Bleed Pool Party shot, Team 2's Russian Naval AI label is mostly invisible to me. Adding the shadow definitely helps me. (Not all color vision deficiencies are equal)

As for the shadows, bottom right is much better than top left. (that is, light coming from top left to cast shadow to bottom right) Team 4 Rush AI in bottom right is a bit hard to read but much easier to read than no shadow at all. (Where it is actually not visible to me. I can tell something is there, but am unable to read it.)

Having written similar code many times I've come to the conclusion that simply redrawing the text offset by 1.4 pixels doesn't always end up being the right answer. For the stuff I'm currently doing, I actually draw a blob and put the text on it. I do use it to solve a slightly different problem. This likely won't fit the UI needs of OpenRA, though.

![image](https://cloud.githubusercontent.com/assets/113401/17116628/f032efb0-5275-11e6-8b5b-78c642f8339a.png)

I do hope this all makes sense. I wrote this rather hastily.

x-a-n-a-x commented:
I am really happy that you decided to implement this shadow and shadow-color, @obrakmann.
I really need to be able to add a shadow on the main menu of the Dune 2 mod (mainmenu.yaml).

Here is the original Dune2 main menu: https://github.com/OpenRA/OpenRA/issues/3274#issuecomment-29591785 (click to see what I want to achieve).

Here is what I can do (with Contrast: true): 
![menu](https://cloud.githubusercontent.com/assets/1732773/17184617/90b94cb4-542d-11e6-9155-ee5fc9861040.png)

obrakmann commented:
Going to add shadow support to the `ButtonWidget` as well then, to satisfy xanax's request. That will make it available on a whole bunch of other widgets that derive from `ButtonWidget`, too.

obrakmann commented:
Updated:

* Added the Shadow option to ButtonWidget as well
* Added shadows to the lobby chat, which I had missed before

x-a-n-a-x commented:
It works great on the main menu :

![menu2](https://cloud.githubusercontent.com/assets/1732773/17270324/3f33df3c-5660-11e6-9b38-8095135c9ae8.png)

pchote commented:
The player bin in the replay browser doesn't switch to the light colored shadow for dark player colors.

pchote commented:
The observer stats doesn't use light shadows for player colors, and can you please also add shadows to the other stats in the rows for consistency?

pchote commented:
The currently selected player name in the observer shroud selector is also missing a shadow.

pchote commented:
We have some mismatches in the chat highlighting:
* In the lobby player names use the default shadow color only
* In the overlay (no text box) ingame chat display player names use the smart alternate color, but displayed as a contrast instead of a shadow
* In the textbox chat display player names don't use any contrast or shadow.

pchote commented:
This is a really nice polish improvement overall.  :+1: after fixing the above nits.

obrakmann commented:
Updated. Changes:

* Fixed the missing shadows on the shroud dropdowns.
* Fixed the TD shadow colors.
* Added shadows to all the columns in the observer stats window.
* Added shadow to the line graph in the observer stats window.
* Added shadow to more chat messages. The chat overlay (ie, the one without visible UI) continues to use the contrast option with automatic color selection, while everything that is drawn on top of UI elements uses shadow with the default colors from metrics.yaml.

I think that's about it. I might still have missed something, so someone give this another once-over, please.

pchote commented:
We'll also need this on the actor tags and "show custom terrain" debug overlays (everyone, including me, seems to have forgotten about these).

obrakmann commented:
Done. The floating text for bounties now also uses it.

Mailaender commented:
Looks a bit blurred to me in cases where it isn't really necessary to add contrast:

![image](https://cloud.githubusercontent.com/assets/756669/17457011/72d4533c-5bea-11e6-9b5b-4c6deb697922.png)

pchote commented:
We could avoid that with per-case checks to only add the shadow if the main color is "too close" to the background (hello, colour validator!)... but that is kind of ugly.

reaperrr commented:
For the record, I agree that bottom-right shadow looks much better (that's what WW used for TS' font, too).

pchote commented:
> Looks a bit blurred to me in cases where it isn't really necessary to add contrast:

In that situation it looks much nicer to use `WidgetUtils.GetContrastColor` which returns black:

![screen shot 2016-08-26 at 18 19 52](https://cloud.githubusercontent.com/assets/167819/18014255/f92609a6-6bb9-11e6-9ba1-d1b22fbc43c0.png)

Unfortunately this then looks quite nasty when you have a dark coloured name:

![screen shot 2016-08-26 at 18 20 21](https://cloud.githubusercontent.com/assets/167819/18014269/06aa514a-6bba-11e6-905e-13f0ea6427c2.png)

pchote commented:
It's more work and boilerplate, but my best suggestion right now to make this look polished in more situations is to change `WidgetUtils.GetContrastColor` to take the light/dark colours as arguments, and then let the mods define those on a per-widget basis.

obrakmann committed:
Split ChromeMetrics' ContrastColor options into two

One for dark backgrounds, one for light ones. The light one isn't currently being used, but will be by future commits

obrakmann committed:
Rewrite color inversal in a reusable way

obrakmann committed:
Add a Shadow option to LabelWidget and ButtonWidget

obrakmann committed:
Add a shadow to text drawn in player colors

obrakmann commented:
OK, updated. Not sure if I'm happy with this.

abcdefg30 commented:
Good work!

#7F7F7F looks still a bit blurred imho, but I can't come up with a better color:
![shadows](https://cloud.githubusercontent.com/assets/7704140/18130673/3d3213ea-6f90-11e6-885c-5d3db9c64b4f.png)

I only ask you to remove the contrast from the chat ([ra chrome](https://github.com/OpenRA/OpenRA/blob/bleed/mods/ra/chrome/ingame-chat.yaml#L17), [cnc chrome](https://github.com/OpenRA/OpenRA/blob/c9e7bb43c2dacee27b66610d4cf9174dc424161f/mods/cnc/chrome/ingame-chat.yaml#L17)),
as it looks horrible with contrast, imho:
![shadowstext](https://cloud.githubusercontent.com/assets/7704140/18131397/f5638834-6f92-11e6-8880-b96aebd96ada.png)

Otherwise 👍 .

pchote commented:
It probably makes sense to replace the contrast on chat and the superweapon timers with shadow.  Yay scopecreep.

obrakmann committed:
Replace use of Contrast with Shadow for ingame chat and support power timers

obrakmann commented:
Updated.

abcdefg30 commented on mods/cnc/chrome/ingame-chat.yaml:
```diff
@@ -14,7 +14,7 @@ Container@CHAT_PANEL:
 					Height: PARENT_BOTTOM
 					RemoveTime: 250
-					UseContrast: yes
+					UseShadow: yes
```
`true` instead of `yes`, please.

pchote commented on OpenRA.Game/Widgets/WidgetUtils.cs:
```diff
@@ -248,6 +248,12 @@ public static string TruncateText(string text, int width, SpriteFont font)
+		public static Color GetContrastColor(Color fgColor, Color bgDark, Color bgLight)
+		{
+			var fg = new HSLColor(fgColor);
+			return fg.RGB == Color.White || fg.L > 128 ? bgDark : bgLight;
```
The shadows/contrast look *much* better if you change this 128 to 64.  This makes it err towards the dark background for the mid-range colors where the light looks bad but dark is perfectly fine.

pchote commented:
> Add contrast to text drawn in player colors

This commit is adding shadows instead of contrast, so please amend this commit message.

pchote commented:
:+1: after those minor tweaks.

obrakmann commented:
Done. I made it 80 instead of 64, though, since it got too dark on dark backgrounds otherwise.

pchote commented:
:+1:, :clap:, and https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/62381e0070eb3481acfc174dc6d26a89e819c737.

Mailaender commented:
This isn't very readable when using `/debugcustomterrain` on Tiberian Sun:

![image](https://cloud.githubusercontent.com/assets/756669/18223694/caf95554-71bf-11e6-8186-0815a5e6347d.png)

Mailaender commented:
This will need either an upgrade rule or some documentation as it causes a rather cryptic error for all thirdparty mods:

```
Exception of type `System.Reflection.TargetInvocationException`: Exception has been thrown by the target of an invocation.
Inner Exception of type `System.Collections.Generic.KeyNotFoundException`: The given key was not present in the dictionary.
    at System.ThrowHelper.ThrowKeyNotFoundException () [0x00000] in /home/abuild/rpmbuild/BUILD/mono-4.4.2/external/referencesource/mscorlib/system/throwhelper.cs:68 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].get_Item (System.Collections.Generic.TKey key) [0x00021] in /home/abuild/rpmbuild/BUILD/mono-4.4.2/external/referencesource/mscorlib/system/collections/generic/dictionary.cs:183 
  at OpenRA.Widgets.ChromeMetrics.Get[T] (System.String key) [0x00008] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/ChromeMetrics.cs:33 
  at OpenRA.Mods.Common.Widgets.LabelWidget..ctor () <0x40a1e470 + 0x0012f> in <filename unknown>:0 
  at (wrapper managed-to-native) System.Reflection.MonoCMethod:InternalInvoke (System.Reflection.MonoCMethod,object,object[],System.Exception&)
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00002] in /home/abuild/rpmbuild/BUILD/mono-4.4.2/mcs/class/corlib/System.Reflection/MonoMethod.cs:644 
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00013] in /home/abuild/rpmbuild/BUILD/mono-4.4.2/mcs/class/corlib/System.Reflection/MonoMethod.cs:650 
  at System.Reflection.MonoCMethod.DoInvoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00089] in /home/abuild/rpmbuild/BUILD/mono-4.4.2/mcs/class/corlib/System.Reflection/MonoMethod.cs:635 
  at System.Reflection.MonoCMethod.Invoke (BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-4.4.2/mcs/class/corlib/System.Reflection/MonoMethod.cs:663 
  at System.Reflection.ConstructorInfo.Invoke (System.Object[] parameters) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-4.4.2/mcs/class/corlib/System.Reflection/ConstructorInfo.cs:62 
  at OpenRA.ObjectCreator.CreateBasic (System.Type type) [0x00013] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/ObjectCreator.cs:126 
  at OpenRA.ObjectCreator.CreateObject[T] (System.String className, System.Collections.Generic.Dictionary`2 args) [0x00050] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/ObjectCreator.cs:103 
  at OpenRA.WidgetLoader.NewWidget (System.String widgetType, OpenRA.Widgets.WidgetArgs args) [0x0002c] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:85 
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, OpenRA.MiniYamlNode node) [0x00033] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:52 
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, OpenRA.MiniYamlNode node) [0x0014d] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:69 
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, System.String w) [0x00032] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:44 
  at OpenRA.Game.LoadWidget (OpenRA.World world, System.String id, OpenRA.Widgets.Widget parent, OpenRA.Widgets.WidgetArgs args) [0x00041] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:136 
  at OpenRA.Mods.Common.Traits.LoadWidgetAtGameStart.WorldLoaded (OpenRA.World world, OpenRA.Graphics.WorldRenderer wr) <0x40a1aed0 + 0x000ab> in <filename unknown>:0 
  at OpenRA.World.LoadComplete (OpenRA.Graphics.WorldRenderer wr) [0x00083] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/World.cs:229 
  at OpenRA.Game.StartGame (System.String mapUID, WorldType type) [0x000c8] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:171 
  at OpenRA.Game.LoadShellMap () [0x00019] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:447 
  at OpenRA.Mods.Common.LoadScreens.BlankLoadScreen.StartGame (OpenRA.Arguments args) <0x40985520 + 0x003b7> in <filename unknown>:0 
  at OpenRA.Game.InitializeMod (System.String mod, OpenRA.Arguments args) [0x0038f] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:434 
  at OpenRA.Game.Initialize (OpenRA.Arguments args) [0x0034c] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:323 
  at OpenRA.Program.Run (System.String[] args) [0x00007] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:117 
  at OpenRA.Program.Main (System.String[] args) [0x00050] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:40 
```

pchote commented:
The best fix would be to add a proper check and exception message inside ChromeMetrics.Get.
